File: Android/tuilivekit/voiceroomcore/src/main/java/com/trtc/uikit/livekit/voiceroomcore/view/RippleView.java
package com.trtc.uikit.livekit.voiceroomcore.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;

import com.tencent.qcloud.tuicore.util.ScreenUtil;
import com.trtc.uikit.livekit.voiceroomcore.R;

public class RippleView extends View {
    private static final int STROKE_WIDTH_DEFAULT = 2;
    private static final int SPEED_DEFAULT        = 1;
    private static final int VIEW_WIDTH_DEFAULT   = 120;
    private static final int MIN_ALPHA            = 40;

    private Paint mPaint;
    private float mWidth;
    private float mHeight;

    private final Circle mCircle;

    private static final int   mSpeed = SPEED_DEFAULT;
    private final        int   mColor;
    private final        float mStartRadius;

    public RippleView(Context context) {
        this(context, null);
    }

    public RippleView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RippleView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        @SuppressLint("CustomViewStyleable")
        TypedArray tya = context.obtainStyledAttributes(attrs, R.styleable.VoiceRoomCoreRippleView);
        mColor = tya.getColor(R.styleable.VoiceRoomCoreRippleView_color, Color.BLUE);
        mStartRadius = tya.getDimension(R.styleable.VoiceRoomCoreRippleView_start_radius, 0f);
        mCircle = new Circle(mStartRadius, 255);
        tya.recycle();
        init();
    }

    private void init() {
        mPaint = new Paint();
        mPaint.setColor(mColor);
        mPaint.setStrokeWidth(ScreenUtil.dip2px(STROKE_WIDTH_DEFAULT));
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeCap(Paint.Cap.ROUND);
        mPaint.setAntiAlias(true);
        setBackgroundColor(Color.TRANSPARENT);
    }

    @Override
    public void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);
        drawCircle(canvas);
    }

    private void drawCircle(Canvas canvas) {
        canvas.save();
        mPaint.setAlpha(mCircle.alpha);
        float strokeWidth = mPaint.getStrokeWidth();
        canvas.drawCircle(mWidth / 2, mHeight / 2, mCircle.width, mPaint);
        if (mCircle.width > (mWidth / 2 - strokeWidth / 2)) {
            mCircle.width = mStartRadius;
            mCircle.alpha = 255;
        } else {
            double alpha = MIN_ALPHA + 255 - mCircle.width * (255 / ((double) (mWidth / 2 + strokeWidth / 2)));
            mCircle.alpha = (int) alpha;
            mCircle.width += mSpeed;
        }

        invalidate();
        canvas.restore();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int myWidthSpecMode = MeasureSpec.getMode(widthMeasureSpec);
        int myWidthSpecSize = MeasureSpec.getSize(widthMeasureSpec);
        int myHeightSpecMode = MeasureSpec.getMode(heightMeasureSpec);
        int myHeightSpecSize = MeasureSpec.getSize(heightMeasureSpec);
        if (myWidthSpecMode == MeasureSpec.EXACTLY) {
            mWidth = myWidthSpecSize;
        } else {
            mWidth = ScreenUtil.dip2px(VIEW_WIDTH_DEFAULT);
        }

        if (myHeightSpecMode == MeasureSpec.EXACTLY) {
            mHeight = myHeightSpecSize;
        } else {
            mHeight = ScreenUtil.dip2px(VIEW_WIDTH_DEFAULT);
        }
        setMeasuredDimension((int) mWidth, (int) mHeight);
    }


    private static class Circle {
        float width;
        int   alpha;

        public Circle(float width, int alpha) {
            this.width = width;
            this.alpha = alpha;
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/voiceroomcore/view/RippleView.kt
package com.trtc.uikit.livekit.voiceroomcore.view

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.trtc.tuikit.common.util.ScreenUtil
import com.trtc.uikit.livekit.R

@SuppressLint("CustomViewStyleable")
class RippleView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        strokeWidth = ScreenUtil.dip2px(2f).toFloat()
        style = Paint.Style.STROKE
        strokeCap = Paint.Cap.ROUND
    }

    private var circleWidth = 0f
    private var circleAlpha = 255
    private var startRadius = 0f

    init {
        context.obtainStyledAttributes(attrs, R.styleable.VoiceRoomCoreRippleView).apply {
            paint.color = getColor(R.styleable.VoiceRoomCoreRippleView_color, Color.BLUE)
            startRadius = getDimension(R.styleable.VoiceRoomCoreRippleView_start_radius, 0f)
            recycle()
        }
        circleWidth = startRadius
        setBackgroundColor(Color.TRANSPARENT)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val defaultSize = ScreenUtil.dip2px(120f)
        setMeasuredDimension(
            resolveSize(defaultSize, widthMeasureSpec),
            resolveSize(defaultSize, heightMeasureSpec)
        )
    }

    override fun onDraw(canvas: Canvas) {
        paint.alpha = circleAlpha
        val centerX = width / 2f
        val centerY = height / 2f
        val maxRadius = centerX - paint.strokeWidth / 2

        canvas.drawCircle(centerX, centerY, circleWidth, paint)

        when {
            circleWidth > maxRadius -> resetCircle()
            else -> updateCircle(maxRadius)
        }
        invalidate()
    }

    private fun resetCircle() {
        circleWidth = startRadius
        circleAlpha = 255
    }

    private fun updateCircle(maxRadius: Float) {
        circleAlpha = (40 + 215 * (1 - circleWidth / maxRadius)).toInt()
        circleWidth += 1
    }
}